include/aspect/boundary_temperature/dynamic_core.h

```diff
@@ -0,0 +1,521 @@
+     * and the core-mantle boundaries (CMB) temperature is calculated by core energy balance.
+     * The formulation of core energy balance are from:
+     * Nimmo et al., The influence of potassium on core and geodynamo evolution.
+     *    Geophysical Journal International, 2004. 156(2): p. 363-376.
+     * @ingroup BoundaryTemperatures
+     */
+    template <int dim>
+    class DynamicCore : public Interface<dim>, public aspect::SimulatorAccess<dim>
+    {
+      public:
+        /**
+         * Construstor
+         */
+        DynamicCore();
+
+        /**
+         * Return the temperature that is to hold at a particular location on the
+         * boundary of the domain. This function returns the temperatures
+         * at the inner and outer boundaries.
+         *
+         * @param boundary_indicator The boundary indicator of the part of the boundary
+         *   of the domain on which the point is located at which we are requesting the
+         *   temperature.
+         * @param location The location of the point at which we ask for the temperature.
+         **/
+        virtual
+        double  boundary_temperature (const types::boundary_id            boundary_indicator,
+                                      const Point<dim>                    &location) const;
+
+        /**
+         * Return the minimal temperature on that part of the boundary
+         * on which Dirichlet conditions are posed.
+         *
+         * This value is used in computing dimensionless numbers such as the
+         * Nusselt number indicating heat flux.
+         */
+        virtual
+        double minimal_temperature (const std::set<types::boundary_id> &fixed_boundary_ids) const;
+
+        /**
+         * Return the maximal temperature on that part of the boundary
+         * on which Dirichlet conditions are posed.
+         *
+         * This value is used in computing dimensionless numbers such as the
+         * Nusselt number indicating heat flux.
+         */
+        virtual
+        double maximal_temperature (const std::set<types::boundary_id> &fixed_boundary_ids) const;
+
+        /**
+         * Declare the parameters this class takes through input files.
+         * This class declares the inner and outer boundary temperatures.
+         */
+        static
+        void
+        declare_parameters (ParameterHandler &prm);
+
+        /**
+         * Read the parameters this class declares from the parameter
+         * file.
+         */
+        virtual
+        void
+        parse_parameters (ParameterHandler &prm);
+
+        /**
+         * This function update the core-mantle boundary (CMB) temperature by
+         * the core energy balance solver using the core-mantle boundary heat flux.
+         */
+        virtual void update();
+
+        /**
+         * Pass core data to other modules
+         */
+        struct internal::CoreData get_core_data() const;
```

Comment:
@bangerth Thanks for your comments. The core data is passed from boundary temperature to postprocess once each time step, and it passed back from postprocess to boundary temperature only at the restart. And if in future, we move the heat flux calculation into postprocess as @gassmoeller proposed, we will need pass data from postprocess to boundary temperature every time step as well.
So it's not a single way data passing. I know you guys don't like keep multiple copy of the same data, it's the only solution I can figure out at the moment. I don't think const reference will work for this case. Even we pass a reference, we still need to manually copy the data over.